### PR TITLE
PhysSim: Different approaches to reach network relaxation

### PR DESCRIPTION
**Pull request by REASY**
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2403)

### CODE REVIEW

File: src/main/scala/beam/agentsim/agents/modalbehaviors/DrivesVehicle.scala [hunk 1]
```diff
@@ -531,7 +533,8 @@ trait DrivesVehicle[T <: DrivingData] extends BeamAgent[T] with Stash {
           fuelConsumed.secondaryFuel,
           currentBeamVehicle.primaryFuelLevelInJoules,
           currentBeamVehicle.secondaryFuelLevelInJoules,
-          tollOnCurrentLeg /*,
+          tollOnCurrentLeg,
+          currentLeg.request /*,
```
Comment (wrashid) [inline src/main/scala/beam/agentsim/agents/modalbehaviors/DrivesVehicle.scala]: why was this changed needed?

File: src/main/scala/beam/sim/BeamHelper.scala [hunk 2]
```diff
@@ -643,7 +653,8 @@ trait BeamHelper extends LazyLogging {
 
     logger.info("Starting beam on branch {} at commit {}.", BashUtils.getBranch, BashUtils.getCommitHash)
 
-    prepareDirectories(config, beamConfig, outputDirectory)
+    if (false)
```
Comment (wrashid) [inline src/main/scala/beam/sim/BeamHelper.scala]: if false?

File: src/main/scala/beam/sim/BeamMobsim.scala [hunk 3]
```diff
@@ -91,6 +94,82 @@ class BeamMobsim @Inject()(
...
+    val experimentType = beamServices.beamConfig.beam.physsim.relaxation.`type`
+    if (experimentType == "experiment_2.0" && beamServices.beamConfig.beam.physsim.relaxation.experiment2_0.clearModesEveryIteration) {
+      clearModes()
+      logger.info(s"Experiment_2.0: Clear all modes at iteration ${iteration}")
+    } else if (experimentType == "experiment_2.1" && beamServices.beamConfig.beam.physsim.relaxation.experiment2_1.clearModesEveryIteration) {
+      clearModes()
+      logger.info(s"Experiment_2.1: Clear all modes at iteration ${iteration}")
+    } else if (experimentType == "experiment_3.0" && iteration <= 1) {
+      clearModes()
+      logger.info(s"Experiment_3.0: Clear all modes at iteration ${iteration}")
```
Comment (wrashid) [inline src/main/scala/beam/sim/BeamMobsim.scala]: would merging clearRoutesIfNeeded and clearModesIfNeeded lead to less code duplication?

File: test/input/beamville/beam.conf [hunk 4]
```diff
@@ -20,7 +26,7 @@ beam.agentsim.endTime = "30:00:00"
 beam.agentsim.agents.modalBehaviors.modeChoiceClass = "ModeChoiceMultinomialLogit"
 beam.agentsim.agents.modalBehaviors.defaultValueOfTime = 8.0
 beam.agentsim.agents.modalBehaviors.mulitnomialLogit.params.transfer = -1.4
-beam.agentsim.agents.modalBehaviors.mulitnomialLogit.params.car_intercept = 2.0
+beam.agentsim.agents.modalBehaviors.mulitnomialLogit.params.car_intercept = 1000.0
```
Comment (wrashid) [inline test/input/beamville/beam.conf]: why?

File: src/main/scala/beam/sim/BeamMobsim.scala (hunk 3 above)
Comment (REASY) [reply src/main/scala/beam/sim/BeamMobsim.scala]: I merged them together to remove code duplication

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 5]
```diff
@@ -111,72 +120,78 @@ public AgentSimToPhysSimPlanConverter(EventsManager eventsManager,
...
+                break;
+            case "experiment_2.1":
+                sim = new Experiment_2_1(beamConfig, agentSimScenario, jdeqsimPopulation,
+                        beamServices,
+                        controlerIO, caccVehiclesMap, beamConfigChangesObservable, iterationNumber, linkStatsGraph, shouldWritePhysSimEvents(iterationNumber), rnd);
+                break;
+            case "experiment_3.0":
+                sim = new Experiment_3_0(beamConfig, agentSimScenario, jdeqsimPopulation,
+                        beamServices,
+                        controlerIO, caccVehiclesMap, beamConfigChangesObservable, iterationNumber, linkStatsGraph, shouldWritePhysSimEvents(iterationNumber), rnd);
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: let us extract to factory?

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 6]
```diff
@@ -398,21 +359,45 @@ public void handleEvent(Event event) {
...
             person.addPlan(plan);
             person.setSelectedPlan(plan);
+            // person.getAttributes().putAttribute("vehicle_type", vehType);
             jdeqsimPopulation.addPerson(person);
+            final Person originalPerson = agentSimScenario.getPopulation().getPersons().get(driverId);
+            final Person personToCopyFrom = originalPerson == null ? agentSimScenario.getPopulation().getPersons().get(vehicleId) : originalPerson;
+            // Try to copy person's attributes from original `agentSimScenario` to the created one. Attributes are important because they are used during R5 routing
+            if (personToCopyFrom != null) {
+                try {
+                    Attributes attributes = personToCopyFrom.getAttributes();
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: are we using attributes here?

File: src/main/resources/beam-template.conf [hunk 7]
```diff
@@ -567,6 +582,7 @@ beam.routing {
     osmMapdbFile = ${beam.inputDirectory}"/r5/osm.mapdb"
     mNetBuilder.fromCRS = "EPSG:4326" # WGS84
     mNetBuilder.toCRS = "EPSG:26910"     # UTM10N
+    travelTimeError = "double | 0.0"
```
Comment (wrashid) [inline src/main/resources/beam-template.conf]: rename: travelTimeNoiseFraction

File: src/main/scala/beam/physsim/jdeqsim/ApproxPhysSim.scala [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+    logger.info(s"Diff in eventTypeToNumberOfMessages map: \n${diffMap.mkString("\n")}")
+    //    PhysSim.printAverageCarTravelTime(getCarPeople)
+    logger.info(s"Car travel time stats at iteration ${prevResult.iteration}: ${prevResult.carTravelTimeStats}")
+    logger.info(s"Car travel time stats at iteration ${currentResult.iteration}: ${currentResult.carTravelTimeStats}")
+  }
+
+  private def getCarPeople(population: Population): Vector[Person] = {
+    val carPeople = population.getPersons.values.asScala
+      .filter(p => !p.getId.toString.contains("bus") && !p.getId.toString.contains(":"))
+      .toVector
```
Comment (wrashid) [inline src/main/scala/beam/physsim/jdeqsim/ApproxPhysSim.scala]: reduce duplication?

File: src/main/scala/beam/physsim/jdeqsim/PhysSimReplayer.scala [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+  def eventsFilter(event: Event): Boolean = {
+    val attribs = event.getAttributes
+    // We need only PathTraversal
+    val isNeededEvent = event.getEventType == "PathTraversal"
+    isNeededEvent
+  }
+
+  def main(args: Array[String]): Unit = {
+    assert(
```
Comment (wrashid) [inline src/main/scala/beam/physsim/jdeqsim/PhysSimReplayer.scala]: is this current?

File: src/main/scala/beam/router/model/BeamLeg.scala [hunk 10]
```diff
@@ -1,7 +1,7 @@
 package beam.router.model
 
 import beam.agentsim.events.SpaceTime
-import beam.router.BeamRouter.Location
+import beam.router.BeamRouter.{Location, RoutingRequest}
```
Comment (wrashid) [inline src/main/scala/beam/router/model/BeamLeg.scala]: roll back

File: src/main/scala/beam/router/r5/R5RoutingWorker.scala [hunk 11]
```diff
@@ -197,7 +204,38 @@ class R5RoutingWorker(workerParams: WorkerParameters) extends Actor with ActorLo
       if (firstMsgTime.isEmpty) firstMsgTime = Some(ZonedDateTime.now(ZoneOffset.UTC))
       val eventualResponse = Future {
         latency("request-router-time", Metrics.RegularLevel) {
-          r5.calcRoute(request)
+          val routeWithError = r5.calcRoute(request)
+//          if (request.departureTime == 24400 && request.streetVehicles.exists(x => x.mode == BeamMode.CAR)) {
+//            val beamLegs = routeWithError.itineraries.flatMap(x => x.legs.map(l => l.beamLeg))
+//            val msg =
+//              s"""Request: ${request}
+//                 |Resp: ${routeWithError}
```
Comment (wrashid) [inline src/main/scala/beam/router/r5/R5RoutingWorker.scala]: remove debug code

File: src/main/scala/beam/utils/analysis/AverageCarTravelTimeAnalysis.scala [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+          val physSimTravelTime = travelTime.getLinkTravelTime(link, currentTime, null, null).ceil.toInt
+          val linkTravelTime = Math.max(physSimTravelTime, minTravelTime)
+          val t = Math.min(linkTravelTime, maxTravelTime)
+          val newDuration = acc + t
+          (newDuration, currentTime + newDuration)
+      }
+      ._1
+  }
+
+  private def calcR5Duration_attempt2(
```
Comment (wrashid) [inline src/main/scala/beam/utils/analysis/AverageCarTravelTimeAnalysis.scala]: remove?

File: src/test/scala/beam/sflight/SfLightRoutePopulationSpec.scala [hunk 13]
```diff
@@ -84,7 +85,7 @@ class SfLightRoutePopulationSpec
                 case BeamTrip(legs) =>
                   legs.map(_.mode) should contain theSameElementsInOrderAs List(WALK)
                   inside(legs.loneElement) {
-                    case BeamLeg(_, mode, _, BeamPath(_, _, _, _, _, _)) =>
+                    case BeamLeg(_, mode, _, BeamPath(_, _, _, _, _, _), _) =>
```
Comment (wrashid) [inline src/test/scala/beam/sflight/SfLightRoutePopulationSpec.scala]: undo

File: test/input/sf-light/urbansim-100k.conf [hunk 14]
```diff
@@ -75,7 +75,7 @@ beam.agentsim.tuning.rideHailPrice = 1.0
 # Physsim
 ###########################
 beam.physsim.inputNetworkFilePath = ${beam.inputDirectory}"/r5/physsim-network.xml"
-beam.physsim.flowCapacityFactor = 0.01
+beam.physsim.flowCapacityFactor = 2.0
```
Comment (wrashid) [inline test/input/sf-light/urbansim-100k.conf]: undo

**Comment by REASY** [comment]
test!

**Comment by REASY** [comment]
test!

**Comment by REASY** [comment]
test!

**Comment by REASY** [comment]
test!
